File: app/src/main/java/com/fanchen/imovie/base/BaseRecyclerFragment.java
package com.fanchen.imovie.base;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.View;
import android.widget.TextView;

import com.fanchen.imovie.R;
import com.fanchen.imovie.db.LiteOrmManager;
import com.fanchen.imovie.entity.JsonSerialize;
import com.fanchen.imovie.entity.face.IHomeRoot;
import com.fanchen.imovie.retrofit.callback.RefreshCallback;
import com.fanchen.imovie.retrofit.RetrofitManager;
import com.fanchen.imovie.retrofit.service.DianxiumeiService;
import com.fanchen.imovie.thread.AsyTaskQueue;
import com.fanchen.imovie.thread.task.AsyTaskListener;
import com.fanchen.imovie.thread.task.AsyTaskListenerImpl;
import com.fanchen.imovie.util.LogUtil;
import com.fanchen.imovie.util.NetworkUtil;
import com.fanchen.imovie.view.CustomEmptyView;
import com.google.gson.Gson;
import com.litesuits.orm.LiteOrm;
import com.litesuits.orm.db.assit.QueryBuilder;
import com.litesuits.orm.db.assit.WhereBuilder;
import com.squareup.picasso.Picasso;

import java.lang.reflect.Type;
import java.util.List;

import butterknife.InjectView;

/**
 *
 * Created by fanchen on 2017/7/22.
 */
public abstract class BaseRecyclerFragment extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener,
        View.OnClickListener, BaseAdapter.OnItemClickListener, BaseAdapter.OnLoadListener {

    @InjectView(R.id.recycle)
    protected RecyclerView mRecyclerView;
    @InjectView(R.id.swipe_refresh_layout)
    protected SwipeRefreshLayout mSwipeRefreshLayout;
    @InjectView(R.id.empty_layout)
    protected CustomEmptyView mCustomEmptyView;
    @InjectView(R.id.tv_recycler_bottom)
    protected TextView mTextView;

    private int page = 1;
    private int pageStart = 1;
    private BaseAdapter mAdapter;
    private Bundle savedInstanceState;

    @Override
    protected int getLayout() {
        return R.layout.fragment_recyclerview;
    }

    @Override
    protected void initFragment(@Nullable Bundle savedInstanceState, Bundle args) {
        super.initFragment(savedInstanceState, args);
        this.savedInstanceState = savedInstanceState;
        setHasOptionsMenu(true);
        TypedValue typedValue = new TypedValue();
        if(activity != null) activity.getTheme().resolveAttribute(R.attr.colorPrimary, typedValue, true);
        if (!checkFieldNull()) {
            mSwipeRefreshLayout.setColorSchemeColors(typedValue.data);
            mSwipeRefreshLayout.setEnabled(hasRefresh());
            mRecyclerView.setLayoutManager(getLayoutManager());
            mRecyclerView.setAdapter(mAdapter = getAdapter(getPicasso()));
        }
        if (useLocalStorage() && savedInstanceState == null) {
            LogUtil.e(BaseRecyclerFragment.class, "尝试加载本地数据");
            loadLocalData(AsyTaskQueue.newInstance());
        } else {
            LogUtil.e(BaseRecyclerFragment.class, "加载网络数据");
            loadData(savedInstanceState, getRetrofitManager(), page = pageStart);
        }
    }

    @Override
    protected void setListener() {
        super.setListener();
        if (!checkFieldNull()) {
            if (hasRefresh())
                mSwipeRefreshLayout.setOnRefreshListener(this);
            mCustomEmptyView.setOnClickListener(this);
            mRecyclerView.addOnScrollListener(scrollListener);
        }
        if (mAdapter != null) {
            if (hasLoad())
                mAdapter.setOnLoadListener(this);
            mAdapter.setOnItemClickListener(this);
        }
    }

    public boolean checkFieldNull() {
        return mSwipeRefreshLayout == null || mCustomEmptyView == null || mRecyclerView == null;
    }

    public RecyclerView getRecyclerView() {
        return mRecyclerView;
    }

    public SwipeRefreshLayout getSwipeRefreshLayout() {
        return mSwipeRefreshLayout;
    }

    /**
     * 能否刷新
     *
     * @return
     */
    protected boolean hasRefresh() {
        return true;
    }

    /**
     * RecyclerView  LayoutManager
     *
     * @return
     */
    public abstract RecyclerView.LayoutManager getLayoutManager();

    /**
     * RecyclerView Adapter
     *
     * @return
     */
    public abstract BaseAdapter getAdapter(Picasso picasso);

    /**
     * 加载网络数据
     *
     * @param page
     */
    public abstract void loadData(Bundle savedInstanceState, RetrofitManager retrofit, int page);

    /**
     * 加载本地缓存的数据
     */
    public void loadLocalData(AsyTaskQueue queue) {

    }

    /**
     * 能否加载更多
     *
     * @return
     */
    protected boolean hasLoad() {
        return true;
    }

    /**
     * 当前页数
     *
     * @return
     */
    public int getPage() {
        return page;
    }

    public void setPageStart(int pageStart) {
        this.pageStart = pageStart;
    }

    /**
     * @return
     */
    public boolean isRefresh() {
        return page == pageStart;
    }

    @Override
    public void onRefresh() {
        loadData(null, getRetrofitManager(), page = pageStart);
    }

    @Override
    public void onLoad() {
        loadData(null, getRetrofitManager(), ++page);
    }

    /**
     * 是否使用本地存储
     *
     * @return
     */
    protected boolean useLocalStorage() {
        return false;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.empty_layout:
                onRefresh();
                break;
        }
    }

    /**
     * @return
     */
    public Type getSerializeClass() {
        return null;
    }

    public int getStaleTime() {
        return 12;
    }

    /**
     * @return
     */
    public String getSerializeKey() {
        return getClass().getSimpleName();
    }

    /**
     *
     */
    private RecyclerView.OnScrollListener scrollListener = new RecyclerView.OnScrollListener() {

        @Override
        public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
            Picasso picasso = getPicasso();
            if (picasso != null) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    picasso.resumeTag(BaseRecyclerFragment.class);
                } else {
                    picasso.pauseTag(BaseRecyclerFragment.class);
                }
            }
        }

    };


    /**
     *
     */
    private class SaveTaskListener extends AsyTaskListenerImpl<Void> {

        private Object response;

        public SaveTaskListener(Object response) {
            this.response = response;
        }

        @Override
        public Void onTaskBackground() {
            if (getLiteOrm() == null) return null;
            //保存key
            LogUtil.e("SaveTaskListener","onTaskBackground key -> " + getSerializeKey());
            getLiteOrm().delete(new WhereBuilder(JsonSerialize.class, "key = ?", new Object[]{getSerializeKey()}));
            getLiteOrm().insert(new JsonSerialize(response, getSerializeKey()));
            return null;
        }

    }

    /**
     * @param <T>
     */
    public abstract class QueryTaskListener<T> extends AsyTaskListenerImpl<T> {

        @Override
        public void onTaskSart() {
            if (mCustomEmptyView == null || mSwipeRefreshLayout == null || mAdapter == null) return;
            mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_NON);
            if (!mSwipeRefreshLayout.isRefreshing() && !mAdapter.isLoading()) {
                mSwipeRefreshLayout.setRefreshing(true);
            }
        }

        @Override
        public T onTaskBackground() {
            if (getLiteOrm() == null) return null;
            List<JsonSerialize> query = getLiteOrm().query(new QueryBuilder<>(JsonSerialize.class).where("key = ?", getSerializeKey()));
            if (query != null && query.size() > 0) {
                JsonSerialize jsonSerialize = query.get(0);
                //数据未过期 或者当前无网络情况下，返回缓存的数据
                if (!jsonSerialize.isStale(getStaleTime()) || !NetworkUtil.isNetWorkAvailable(activity)) {
                    try {
                        Type serializeClass = getSerializeClass();
                        if (serializeClass != null) {
                            return (T) new Gson().fromJson(jsonSerialize.getJson(), serializeClass);
                        } else if (jsonSerialize.isRawType()) {
                            Class<?> forName = Class.forName(jsonSerialize.getClazz());
                            return (T) new Gson().fromJson(jsonSerialize.getJson(), forName);
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
            return null;
        }

        @Override
        public void onTaskSuccess(T data) {
            if (mSwipeRefreshLayout == null) return;
            if (data != null) {
                onSuccess(data);
                mSwipeRefreshLayout.setRefreshing(false);
                LogUtil.e(BaseRecyclerFragment.class, "加载本地数据");
            } else {
                //加载网络数据
                LogUtil.e(BaseRecyclerFragment.class, "加载网络数据");
                loadData(savedInstanceState, getRetrofitManager(), page = pageStart);
            }
        }

        /**
         * @param date
         */
        public abstract void onSuccess(T date);
    }


    /**
     * @param <T>
     */
    protected abstract class RefreshRecyclerFragmentImpl<T> implements RefreshCallback<T> {

        @Override
        public void onStart(int enqueueKey) {
            if (mCustomEmptyView == null || mSwipeRefreshLayout == null) return;
            mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_NON);
            if (!mSwipeRefreshLayout.isRefreshing() && !mAdapter.isLoading()) {
                mSwipeRefreshLayout.setRefreshing(true);
            }
        }

        @Override
        public void onFailure(int enqueueKey, String throwable) {
            if (mCustomEmptyView == null || mAdapter == null) return;
            if (mAdapter.isEmpty()) mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_ERROR);
            showSnackbar(throwable);
        }

        @Override
        public void onFinish(int enqueueKey) {
            if (mCustomEmptyView == null || mAdapter == null) return;
            mSwipeRefreshLayout.setRefreshing(false);
            mAdapter.setLoading(false);
            if (mAdapter.isEmpty()) {
                mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_EMPTY);
            }
        }

        @Override
        public void onSuccess(int enqueueKey, T response) {
            if (response == null || mAdapter == null) return;
            if (useLocalStorage()) {
                //将数据序列化到本地
                LogUtil.e(BaseRecyclerFragment.class, "序列化数据到本地" + new Gson().toJson(response));
                AsyTaskQueue.newInstance().execute(new SaveTaskListener(response));
            }
            onSuccess(response,isRefresh());
        }

        /**
         * @param response
         */
        public abstract void onSuccess(T response,boolean refresh);
    }

    protected abstract class TaskRecyclerFragmentImpl<T> implements AsyTaskListener<T> {

        @Override
        public void onTaskFinish() {
            if (isDetached() || !isAdded()) return;
            mSwipeRefreshLayout.setRefreshing(false);
            mAdapter.setLoading(false);
            if (mAdapter.isEmpty()) {
                mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_EMPTY);
            }
        }

        @Override
        public void onTaskSart() {
            if (isDetached() || !isAdded()) return;
            mCustomEmptyView.setEmptyType(CustomEmptyView.TYPE_NON);
            if (!mSwipeRefreshLayout.isRefreshing() && !mAdapter.isLoading()) {
                mSwipeRefreshLayout.setRefreshing(true);
            }
        }

        @Override
        public void onTaskProgress(Integer... values) {
        }

    }

}

File: app/src/main/java/com/fanchen/imovie/base/BaseRecyclerFragment.kt
package com.fanchen.imovie.base

import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.fanchen.imovie.R
import com.fanchen.imovie.entity.JsonSerialize
import com.fanchen.imovie.retrofit.RetrofitManager
import com.fanchen.imovie.retrofit.callback.RefreshCallback
import com.fanchen.imovie.thread.AsyTaskQueue
import com.fanchen.imovie.thread.task.AsyTaskListener
import com.fanchen.imovie.thread.task.AsyTaskListenerImpl
import com.fanchen.imovie.util.LogUtil
import com.fanchen.imovie.util.NetworkUtil
import com.fanchen.imovie.view.CustomEmptyView
import com.google.gson.Gson
import com.litesuits.orm.db.assit.QueryBuilder
import com.litesuits.orm.db.assit.WhereBuilder
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_recyclerview.*
import kotlinx.android.synthetic.main.fragment_recyclerview.view.*
import java.lang.reflect.Type

abstract class BaseRecyclerFragment : BaseFragment(), SwipeRefreshLayout.OnRefreshListener, View.OnClickListener, BaseAdapter.OnItemClickListener, BaseAdapter.OnLoadListener {
    /**
     * 当前页数
     *
     * @return
     */
    var page = 1
    private var pageStart = 1
    private var mAdapter: BaseAdapter? = null
    private var savedInstanceState: Bundle? = null

    /**
     * RecyclerView  LayoutManager
     *
     * @return
     */
    abstract val layoutManager: RecyclerView.LayoutManager

    /**
     * @return
     */
    val isRefresh: Boolean
        get() = page == pageStart

    /**
     * @return
     */
    open val serializeClass: Type?
        get() = null

    open val staleTime: Int
        get() = 12

    /**
     * @return
     */
    open val serializeKey: String
        get() = javaClass.simpleName

    lateinit var mTextView: TextView
    var mRecyclerView: RecyclerView? = null
    var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    lateinit var mCustomEmptyView: CustomEmptyView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        mTextView = view!!.tv_recycler_bottom
        mRecyclerView = view.recycle
        mSwipeRefreshLayout = view.swipe_refresh_layout
        mCustomEmptyView = view.empty_layout
        return view
    }

    fun getSwipeRefreshLayout(): SwipeRefreshLayout? {
        return mSwipeRefreshLayout
    }

    fun getRecyclerView(): RecyclerView? {
        return mRecyclerView
    }

    /**
     *
     */
    private val scrollListener = object : RecyclerView.OnScrollListener() {

        override fun onScrollStateChanged(recycle: RecyclerView?, newState: Int) {
            val picasso = picasso
            if (picasso != null) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    picasso.resumeTag(BaseRecyclerFragment::class.java)
                } else {
                    picasso.pauseTag(BaseRecyclerFragment::class.java)
                }
            }
        }

    }

    override fun getLayout(): Int {
        return R.layout.fragment_recyclerview
    }

    override fun initFragment(savedInstanceState: Bundle?, args: Bundle?) {
        super.initFragment(savedInstanceState, args)
        this.savedInstanceState = savedInstanceState
        setHasOptionsMenu(true)
        val typedValue = TypedValue()
        if (activity != null) activity.theme.resolveAttribute(R.attr.colorPrimary, typedValue, true)
        if (!checkFieldNull()) {
            swipe_refresh_layout!!.setColorSchemeColors(typedValue.data)
            swipe_refresh_layout!!.isEnabled = hasRefresh()
            recycle.layoutManager = layoutManager
            recycle.adapter = getAdapter(picasso)
        }
        if (useLocalStorage() && savedInstanceState == null) {
            loadLocalData(AsyTaskQueue.newInstance())
        } else {
            LogUtil.e(BaseRecyclerFragment::class.java, "加载网络数据")
            loadData(savedInstanceState, retrofitManager, page = pageStart)
        }
    }

    override fun setListener() {
        super.setListener()
        if (!checkFieldNull()) {
            if (hasRefresh())
                swipe_refresh_layout!!.setOnRefreshListener(this)
            empty_layout!!.setOnClickListener(this)
            recycle!!.addOnScrollListener(scrollListener)
        }
        if (mAdapter != null) {
            if (hasLoad())
                mAdapter!!.setOnLoadListener(this)
            mAdapter!!.setOnItemClickListener(this)
        }
    }

    private fun checkFieldNull(): Boolean {
        return swipe_refresh_layout == null || empty_layout == null || recycle == null
    }

    /**
     * 能否刷新
     *
     * @return
     */
    protected open fun hasRefresh(): Boolean {
        return true
    }

    /**
     * RecyclerView Adapter
     *
     * @return
     */
    abstract fun getAdapter(picasso: Picasso?): BaseAdapter

    /**
     * 加载网络数据
     *
     * @param page
     */
    abstract fun loadData(savedInstanceState: Bundle?, retrofit: RetrofitManager?, page: Int)

    /**
     * 加载本地缓存的数据
     */
    open fun loadLocalData(queue: AsyTaskQueue) {

    }

    /**
     * 能否加载更多
     *
     * @return
     */
    protected open fun hasLoad(): Boolean {
        return true
    }

    fun setPageStart(pageStart: Int) {
        this.pageStart = pageStart
    }

    override fun onRefresh() {
        loadData(null, retrofitManager, page = pageStart)
    }

    override fun onLoad() {
        loadData(null, retrofitManager, ++page)
    }

    /**
     * 是否使用本地存储
     *
     * @return
     */
    protected open fun useLocalStorage(): Boolean {
        return false
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.empty_layout -> onRefresh()
        }
    }


    /**
     *
     */
    private inner class SaveTaskListener(private val response: Any) : AsyTaskListenerImpl<Void>() {

        override fun onTaskBackground(): Void? {
            if (liteOrm == null) return null
            //保存key
            liteOrm!!.delete(WhereBuilder(JsonSerialize::class.java, "key = ?", arrayOf<Any>(serializeKey)))
            liteOrm!!.insert(JsonSerialize(response, serializeKey))
            return null
        }

    }

    /**
     * @param <T>
    </T> */
    abstract inner class QueryTaskListener<T> : AsyTaskListenerImpl<T>() {

        override fun onTaskStart() {
            if (empty_layout == null || swipe_refresh_layout == null || mAdapter == null) return
            empty_layout!!.setEmptyType(CustomEmptyView.TYPE_NON)
            if (!swipe_refresh_layout!!.isRefreshing && !mAdapter!!.isLoading) {
                swipe_refresh_layout!!.isRefreshing = true
            }
        }

        override fun onTaskBackground(): T? {
            if (liteOrm == null) return null
            val query = liteOrm!!.query(QueryBuilder(JsonSerialize::class.java).where("key = ?", serializeKey))
            if (query != null && query.size > 0) {
                val jsonSerialize = query[0]
                //数据未过期 或者当前无网络情况下，返回缓存的数据
                if (!jsonSerialize.isStale(staleTime) || !NetworkUtil.isNetWorkAvailable(activity)) {
                    try {
                        val serializeClass = serializeClass
                        if (serializeClass != null) {
                            return Gson().fromJson<T>(jsonSerialize.json, serializeClass)
                        } else if (jsonSerialize.isRawType) {
                            val forName = Class.forName(jsonSerialize.clazz)
                            return Gson().fromJson<T>(jsonSerialize.json, forName)
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
            }
            return null
        }

        override fun onTaskSuccess(data: T?) {
            if (swipe_refresh_layout == null) return
            if (data != null) {
                onSuccess(data)
                swipe_refresh_layout!!.isRefreshing = false
                LogUtil.e(BaseRecyclerFragment::class.java, "加载本地数据")
            } else {
                //加载网络数据
                LogUtil.e(BaseRecyclerFragment::class.java, "加载网络数据")
                loadData(savedInstanceState, retrofitManager, page = pageStart)
            }
        }

        /**
         * @param date
         */
        abstract fun onSuccess(date: T)
    }


    /**
     * @param <T>
    </T> */
    protected abstract inner class RefreshRecyclerFragmentImpl<T> : RefreshCallback<T> {

        override fun onStart(enqueueKey: Int) {
            if (empty_layout == null || swipe_refresh_layout == null) return
            empty_layout!!.setEmptyType(CustomEmptyView.TYPE_NON)
            if (!swipe_refresh_layout!!.isRefreshing && !mAdapter!!.isLoading) {
                swipe_refresh_layout!!.isRefreshing = true
            }
        }

        override fun onFailure(enqueueKey: Int, throwable: String) {
            if (empty_layout == null || mAdapter == null) return
            if (mAdapter!!.isEmpty) empty_layout!!.setEmptyType(CustomEmptyView.TYPE_ERROR)
            showSnackbar(throwable)
        }

        override fun onFinish(enqueueKey: Int) {
            if (empty_layout == null || mAdapter == null || mAdapter == null) return
            swipe_refresh_layout!!.isRefreshing = false
            mAdapter!!.isLoading = false
            if (mAdapter!!.isEmpty) {
                empty_layout!!.setEmptyType(CustomEmptyView.TYPE_EMPTY)
            }
        }

        override fun onSuccess(enqueueKey: Int, response: T?) {
            if (response == null || mAdapter == null) return
            if (isRefresh) mAdapter!!.clear()
            if (useLocalStorage()) {
                //将数据序列化到本地
                LogUtil.e(BaseRecyclerFragment::class.java, "序列化数据到本地")
                AsyTaskQueue.newInstance().execute(SaveTaskListener(response))
            }
            onSuccess(response)
        }

        /**
         * @param response
         */
        abstract fun onSuccess(response: T)
    }

    protected abstract inner class TaskRecyclerFragmentImpl<T> : AsyTaskListener<T> {

        override fun onTaskFinish() {
            if (isDetached || !isAdded) return
            swipe_refresh_layout!!.isRefreshing = false
            mAdapter!!.isLoading = false
            if (mAdapter!!.isEmpty) {
                empty_layout!!.setEmptyType(CustomEmptyView.TYPE_EMPTY)
            }
        }

        override fun onTaskStart() {
            if (isDetached || !isAdded) return
            empty_layout!!.setEmptyType(CustomEmptyView.TYPE_NON)
            if (swipe_refresh_layout != null && !swipe_refresh_layout.isRefreshing && !mAdapter!!.isLoading) {
                swipe_refresh_layout.isRefreshing = true
            }
        }
    }

}
